File: 3661-maximum-walls-destroyed-by-robots/solution.py
import bisect

class Solution:
    def maximumWalls(self, robots: list[int], distance: list[int], walls: list[int]) -> int:
        n = len(robots)
        # Pair robots with their max bullet distances and sort by position
        r_dist = sorted(zip(robots, distance))
        walls.sort()

        def count_range(l, r):
            """Returns the number of unique walls in the inclusive range [l, r]."""
            if l > r:
                return 0
            return bisect.bisect_right(walls, r) - bisect.bisect_left(walls, l)

        # Pre-identify walls located exactly at robot positions
        # These are always destroyed because robot bullet covers its own position
        U = []
        for r, _ in r_dist:
            idx = bisect.bisect_left(walls, r)
            if idx < len(walls) and walls[idx] == r:
                U.append(1)
            else:
                U.append(0)

        # dp0: Max walls using first i robots, with robot i shooting Left
        # dp1: Max walls using first i robots, with robot i shooting Right
        dp0 = count_range(r_dist[0][0] - r_dist[0][1], r_dist[0][0] - 1) + U[0]
        dp1 = U[0]

        for i in range(1, n):
            r_curr, d_curr = r_dist[i]
            r_prev, d_prev = r_dist[i-1]

            # Walls in the gap (r_prev, r_curr)
            # subL_cnt: walls hit by current robot shooting Left
            l_start, l_end = max(r_curr - d_curr, r_prev + 1), r_curr - 1
            subL_cnt = count_range(l_start, l_end)

            # subR_prev_cnt: walls hit by previous robot shooting Right
            r_start_p, r_end_p = r_prev + 1, min(r_prev + d_prev, r_curr - 1)
            subR_prev_cnt = count_range(r_start_p, r_end_p)

            # f_RL: Unique walls hit if prev shoots Right AND curr shoots Left
            if r_start_p > r_end_p:
                f_RL = subL_cnt
            elif l_start > l_end:
                f_RL = subR_prev_cnt
            elif l_start <= r_end_p:
                # Ranges overlap or touch, union is [r_start_p, l_end]
                f_RL = count_range(r_start_p, l_end)
            else:
                # Ranges are disjoint within the gap
                f_RL = subL_cnt + subR_prev_cnt

            next_dp0 = max(dp0 + subL_cnt, dp1 + f_RL) + U[i]
            next_dp1 = max(dp0 + 0, dp1 + subR_prev_cnt) + U[i]
            
            dp0, dp1 = next_dp0, next_dp1

        # The last robot might shoot right into the infinite space beyond it
        last_right_gap = count_range(r_dist[-1][0] + 1, r_dist[-1][0] + r_dist[-1][1])
        
        return int(max(dp0, dp1 + last_right_gap))

File: 3661-maximum-walls-destroyed-by-robots/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maximumWalls(vector<int>& robots, vector<int>& distance, vector<int>& walls) {
        int n = robots.size();
        vector<pair<int, int>> r(n);
        for (int i = 0; i < n; ++i) {
            r[i] = {robots[i], distance[i]};
        }
        sort(r.begin(), r.end());
        sort(walls.begin(), walls.end());

        auto countInRange = [&](int low, int high) {
            if (low > high) return 0;
            auto it1 = lower_bound(walls.begin(), walls.end(), low);
            auto it2 = upper_bound(walls.begin(), walls.end(), high);
            return (int)distance(it1, it2);
        };

        // Walls at robot positions are always destroyed.
        int wallsAtRobots = 0;
        for (int i = 0; i < n; ++i) {
            if (binary_search(walls.begin(), walls.end(), r[i].first)) {
                wallsAtRobots++;
            }
        }

        // dp[i][0]: robot i fires Left
        // dp[i][1]: robot i fires Right
        vector<vector<int>> dp(n, vector<int>(2, 0));

        // Base case: first robot
        dp[0][0] = countInRange(r[0].first - r[0].second, r[0].first - 1);
        dp[0][1] = 0;

        for (int i = 0; i < n - 1; ++i) {
            int r_idx = r[i].first;
            int r_dist = r[i].second;
            int next_r_idx = r[i+1].first;
            int next_r_dist = r[i+1].second;

            // Walls in the gap (R_i, R_{i+1})
            int gap_low = r_idx + 1;
            int gap_high = next_r_idx - 1;

            // Only robot i+1 fires into gap (Left)
            int wL = countInRange(max(next_r_idx - next_r_dist, gap_low), gap_high);
            // Only robot i fires into gap (Right)
            int wR = countInRange(gap_low, min(r_idx + r_dist, gap_high));
            
            // Both fire into gap: Union of ranges
            int intersect_low = max(next_r_idx - next_r_dist, gap_low);
            int intersect_high = min(r_idx + r_dist, gap_high);
            int wBoth = wL + wR - countInRange(intersect_low, intersect_high);

            // dp[i+1][0] (next fires Left)
            dp[i+1][0] = max(dp[i][0] + wL, dp[i][1] + wBoth);
            // dp[i+1][1] (next fires Right)
            dp[i+1][1] = max(dp[i][0] + 0, dp[i][1] + wR);
        }

        int last_r_idx = r[n-1].first;
        int last_r_dist = r[n-1].second;
        int last_right_hit = countInRange(last_r_idx + 1, last_r_idx + last_r_dist);

        return wallsAtRobots + max(dp[n-1][0], dp[n-1][1] + last_right_hit);
    }
};